services/src/main/scala/cromwell/pipeline/service/ProjectSearchService.scala

```diff
@@ -0,0 +1,115 @@
+package cromwell.pipeline.service
+
+import cromwell.pipeline.datastorage.dto._
+import cromwell.pipeline.model.wrapper.UserId
+import cromwell.pipeline.service.ProjectConfigurationService.Exceptions.NotFound
+import cromwell.pipeline.service.ProjectSearchService.Exceptions._
+import cromwell.pipeline.service.exceptions.ServiceException
+
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success }
+
+trait ProjectSearchService {
+
+  def searchProjects(request: ProjectSearchRequest, userId: UserId): Future[Seq[Project]]
+
+}
+
+object ProjectSearchService {
+
+  object Exceptions {
+    sealed abstract class ProjectSearchException(message: String) extends ServiceException(message)
+    final case class InternalError(message: String = "Internal error") extends ProjectSearchException(message)
+  }
+
+  def apply(
+    projectService: ProjectService,
+    projectFileService: ProjectFileService,
+    projectConfigurationService: ProjectConfigurationService
+  )(
+    implicit ec: ExecutionContext
+  ): ProjectSearchService =
+    new ProjectSearchService {
+
+      override def searchProjects(request: ProjectSearchRequest, userId: UserId): Future[Seq[Project]] = {
+        val userProjects = projectService.getUserProjects(userId).recoverWith {
+          case _ => internalError("fetch projects")
+        }
+
+        def doSearch(searchFilter: ProjectSearchFilter): Future[Seq[Project]] = searchFilter match {
+          case AllFilter                    => userProjects
+          case ByNameFilter(mode, value)    => searchWithNameFilter(mode, value, userProjects)
+          case ByFilesFilter(_, hasFiles)   => searchByFilesFilter(hasFiles, userProjects, userId)
+          case ByConfigFilter(_, hasConfig) => searchByConfigFilter(hasConfig, userProjects, userId)
+          case ComplexOrFilter(leftFilter, rightFilter) =>
+            doSearch(leftFilter).flatMap(
+              resultLeft => doSearch(rightFilter).map(resultRight => resultLeft ++ resultRight)
+            )
+          case ComplexAndFilter(leftFilter, rightFilter) =>
+            doSearch(leftFilter).flatMap(
+              resultLeft => doSearch(rightFilter).map(resultRight => resultLeft.intersect(resultRight))
+            )
```

Comment:
Why do we need to traverse all user projects with left filter, then traverse _the same_ projects with right filter and then search for intersection? Can't we feed to right filter the output of the left filter? That will give us their intersection for free and might be much more efficient

Reply:
Refactored